pkg/resto/resto.go

```diff
@@ -0,0 +1,349 @@
+		}
+		res := make(map[string]string)
+		err = json.Unmarshal([]byte(ma.Payload), &res)
+		return res, err
+	}
+	// legacy image mode
+	refs := manifest.References()
+	if len(refs) != 2 {
+		return nil, fmt.Errorf("expected 2 references, found %v", len(refs))
+	}
+	// assume second element is the layer (first being the image config)
+	r := refs[1]
+	rdgst := r.Digest
+	blobsService := repo.Blobs(ctx)
+	payloadGz, err := blobsService.Get(ctx, rdgst)
+	if err != nil {
+		return nil, err
+	}
+	payloadBuf := bytes.NewBuffer(payloadGz)
+	gzf, err := gzip.NewReader(payloadBuf)
+	if err != nil {
+		return nil, err
+	}
+	tarReader := tar.NewReader(gzf)
+	return tarContent(tarReader)
+}
+
+func tarContent(tarReader *tar.Reader) (map[string]string, error) {
+	res := make(map[string]string)
+	for {
+		header, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return res, err
+		}
+		if header.Typeflag == tar.TypeReg {
+			content := bytes.NewBuffer(nil)
+			io.Copy(content, tarReader)
+			res[header.Name] = content.String()
+		}
+	}
+	return res, nil
+}
+
+// PushConfig pushes a configuration file to a registry and returns its digest
+func PushConfig(ctx context.Context, payload, repoTag string, opts RegistryOptions, labels map[string]string) (string, error) {
+	return PushConfigMulti(ctx, map[string]string{
+		"config": payload,
+	}, repoTag, opts, labels)
+}
+
+// PushConfigMulti pushes a set of configuration files to a registry and returns its digest
+func PushConfigMulti(ctx context.Context, payload map[string]string, repoTag string, opts RegistryOptions, labels map[string]string) (string, error) {
+	pr, err := parseRef(repoTag)
+	if err != nil {
+		return "", err
+	}
+	if opts.Username == "" {
+		opts.Username, opts.Password, _ = getCredentials(pr.domain)
+	}
+	repo, err := NewRepository(ctx, pr.domain, pr.path, opts)
+	if err != nil {
+		return "", err
+	}
+	digest, err := pushConfigMediaType(ctx, payload, pr, repo)
+	if err == nil {
+		return digest, err
+	}
+	if _, ok := err.(unsupportedMediaType); ok {
+		return pushConfigLegacy(ctx, payload, pr, repo, labels)
+	}
+	return digest, err
+}
+
+func pushConfigMediaType(ctx context.Context, payload map[string]string, pr parsedReference, repo distribution.Repository) (string, error) {
+	j, err := json.Marshal(payload)
+	if err != nil {
+		return "", err
+	}
+	manifestAny := ManifestAny{
+		Versioned: manifest.Versioned{
+			SchemaVersion: 2,
+			MediaType:     MediaTypeConfig,
+		},
+		Payload: string(j),
+	}
+	raw, err := json.Marshal(manifestAny)
+	if err != nil {
+		return "", err
+	}
+	manifestService, err := repo.Manifests(ctx)
+	if err != nil {
+		return "", err
+	}
+	manifest := NewConfigManifest(MediaTypeConfig, raw)
+	dgst, err := manifestService.Put(ctx, manifest, distribution.WithTag(pr.tag))
+	if err == nil {
+		return dgst.String(), nil
+	}
+	if !strings.Contains(err.Error(), "manifest invalid") && !strings.Contains(err.Error(), "manifest Unknown") {
```

Comment:
Invert the conditions? It will be more readable:
```go
if strings.Contains(err.Error(), "manifest invalid") || strings.Contains(err.Error(), "manifest Unknown") {
    return "", unsupportedMediaType{}
}
return "", err
```
Or even a switch could be nice:
```go
switch{
case strings.Contains(err.Error(), "manifest invalid"):
    return "", unsupportedMediaType{}
case strings.Contains(err.Error(), "manifest Unknown"):
    return "", unsupportedMediaType{}
default:
    return "", err
}
```